Manually add target-aware TLI pass to start of middle-end passes

trenouf opened the pull request:
This stops LLVM's middle-end instcombining from using library functions
    that we don't have. Thus this commit also removes the .ll file
    implementations of tan and ldexpf that are no longer needed.
    
    I had to manually re-enable use of memcpy and memset. See TODO comment
    in the code.
    
    I had to manually disallow use of tan. This can be removed once we have
    LLVM fix https://reviews.llvm.org/D67406
    
    Change-Id: I8a054d819cb7d2368ede89945b79eb41e6321b49

amdrexu commented on context/llpcCompiler.cpp:
```diff
@@ -1325,6 +1326,26 @@ Result Compiler::BuildPipelineInternal(
+        // test/shaderdb/ObjStorageBlock_TestMemCpyInt32.comp
+        // got unrolled far too much, and at too late a stage for the descriptor loads to be commoned up. It might
+        // be an unfortunate interaction between LoopIdiomRecognize and fat pointer laundering.
+        TargetLibraryInfoImpl tli(Triple(pPipelineModule->getTargetTriple()));
```
tli -> targetLibInfo

amdrexu commented on context/llpcCompiler.cpp:
```diff
@@ -1325,6 +1326,26 @@ Result Compiler::BuildPipelineInternal(
+        tli.setUnavailable(LibFunc_tanf);
+        tli.setUnavailable(LibFunc_tanl);
+
+        auto pTliPass = new TargetLibraryInfoWrapperPass(tli);
```
pTliPass -> pTargetLibInfoPass

trenouf commented:
V2: Rebased on reinstated builder trig change; this change now removes
        the llpc.tan ll functions.
    V3: Also add to lowering pass manager, as it also does instcombine.
        Addressed review comments.

amdvlk-admin commented:
### Test summary for commit 4ef02b3c9bd60151b56f0d5e981da68dee41e59f
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-mainline 3523df4d6717c622cf627a7b36233333730deb09</li>
<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details><summary>Build Failed</summary><ul>


</ul></details>

trenouf commented:
retest this please

amdvlk-admin commented:
### Test summary for commit 4ef02b3c9bd60151b56f0d5e981da68dee41e59f
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-mainline 3523df4d6717c622cf627a7b36233333730deb09</li>
<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details><summary>Build Failed</summary><ul>


</ul></details>

trenouf commented:
Rebased on #232

amdvlk-admin commented:
### Test summary for commit c7625ca6d8fd88d03aedc6f271d69bea28e3d48b
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-mainline 3523df4d6717c622cf627a7b36233333730deb09</li>
<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

trenouf commented:
V4: Test fixup, which demonstrates that this commit temporarily loses
        the optimization that pow(2.0,exp) uses llvm.exp2 rather than
        llvm.pow.
    V5: Moved AddTargetLibInfo for lowering passes into Lower::AddPasses; I
        had not noticed that this happens in two different places in
        Compiler. AddTargetLibInfo itself is now in llpcInternal.

amdrexu commented on util/llpcInternal.cpp:
```diff
@@ -749,4 +749,30 @@ bool IsIsaText(
+// Manually add a target-aware TLI pass, so middle-end optimizations do not think that we have library functions.
+void AddTargetLibInfo(
+    legacy::PassManager*  pPassMgr,   // [in/out] Pass manager
+    Context*              pContext)   // [in] LLPC context
```
Could you move up pContext as the first parameter, following this order

pContext,
pPassMgr

We often follow this order: input parameters, inout parameters, output parameters. Parameter such as pContext, pModule, pEntryPoint are often placed as the first parameter.

amdvlk-admin commented:
### Test summary for commit 09dac9538fea853b3b4120eaada43941dfcbdc1c
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details><summary>Build Failed</summary><ul>

/jenkins/workspace/open-source/llpc-github-pr/driver_build/drivers/pal/src/util/lnx/lnxHashProvider.h:36:10: fatal error: openssl/sha.h: No such file or directory

</ul></details>

trenouf commented:
V6: Addressed review comment.

amdrexu commented:
It is good now.

amdvlk-admin commented:
### Test summary for commit abc9f0e4946f3abc43ecd889ef431bb0929eb823
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>
